### PR TITLE
recursion assignment

### PR DESCRIPTION
**Pull request by HagaiNuriel**

### CODE REVIEW

File: src/main/scala/recfun/RecFun.scala [hunk 1]
```diff
@@ -15,16 +17,47 @@ object RecFun extends RecFunInterface {
    * Exercise 1
    */
   def pascal(c: Int, r: Int): Int = {
+//    if (c == 0 || c == r) 1
+//    else pascal(c-1, r-1) + pascal(c, r-1)
```
Comment (maryum375) [inline src/main/scala/recfun/RecFun.scala:21]: Remove commented out code

File: src/main/scala/recfun/RecFun.scala [hunk 2]
```diff
@@ -15,16 +17,47 @@ object RecFun extends RecFunInterface {
...
+    @tailrec
+    def pascalAux(currCol: Int, currRow: Int, prevTriangleRow: Array[Int], currTriangleRow: Array[Int]): Int = {
+      currTriangleRow(currCol) = (if (currCol > 0) prevTriangleRow(currCol - 1) else 0) + (if (currCol < currRow) prevTriangleRow(currCol) else 0)
+
+      if ((currCol == c) && (currRow == r)) currTriangleRow(currCol)
+      else if (currCol < currRow) pascalAux(currCol + 1, currRow, prevTriangleRow, currTriangleRow)
+      else pascalAux(0, currRow + 1, currTriangleRow, new Array(_length = currRow + 2))
+    }
```
Comment (maryum375) [inline src/main/scala/recfun/RecFun.scala:29]: Over complicated solution, try to find a simpler solution.
Generally most of the tasks of coursera can be implemented in a short, smart and efficient way

File: src/main/scala/recfun/RecFun.scala [hunk 3]
```diff
@@ -15,16 +17,47 @@ object RecFun extends RecFunInterface {
...
   /**
    * Exercise 2
    */
-  def balance(chars: List[Char]): Boolean = ???
+  def balance(chars: List[Char]): Boolean = {
+    def balanceRec(chars: List[Char], openParentheses: Int): Boolean = {
+      if (openParentheses < 0) false
+      else if (chars.isEmpty) openParentheses == 0
+      else {
+        val c = chars.head
```
Comment (maryum375) [inline src/main/scala/recfun/RecFun.scala:43]: Personal preference: Use meaningful names for variables. `c` makes it a bit hard to read - especially after some time or some new developer

File: src/main/scala/recfun/RecFun.scala [hunk 4]
```diff
@@ -15,16 +17,47 @@ object RecFun extends RecFunInterface {
...
+        val currOpenParentheses =
+          if (c == '(') openParentheses + 1
+          else if (c == ')') openParentheses - 1
+          else openParentheses
+        balanceRec(chars.tail, currOpenParentheses)
+      }
+    }
+
+    balanceRec(chars, 0)
+  }
```
Comment (maryum375) [inline src/main/scala/recfun/RecFun.scala:53]: Good solution. Simple and clear!

**Review by maryum375 — CHANGES_REQUESTED** [review]

File: src/main/scala/recfun/RecFun.scala (hunk 2 above)
Comment (HagaiNuriel) [reply src/main/scala/recfun/RecFun.scala:29]: if you look at the commented code which I left on purpose, that is my simple clean solution. only problem is that it is not tail recursion :)

Comment (HagaiNuriel) [reply src/main/scala/recfun/RecFun.scala:29]: if you have a shorter solution for a tail recursive implementation of this problem, I'd be happy to see!
What I did here is build (as the "accumulator" of the tail recursion) the last and current rows of the pascal triangle in order compute the current iteration's row and column

File: src/main/scala/recfun/RecFun.scala (hunk 3 above)
Comment (HagaiNuriel) [reply src/main/scala/recfun/RecFun.scala:43]: agree, thanks

**Review by maryum375 — APPROVED** [review]

**Pull request merged**